# **Mini_Project_3 Harshit Arora DS/AIML BATCH_1**

#Business Problem
A Romania -based credit card issuer, Viva-Credit, (https://vivacredit.ro) wants to better predict
the likelihood of default for its customers, as well as identify the key drivers that determine this
likelihood.
This would inform the issuer’s decisions on who to give a credit card to and what credit limit to
provide.
It would also help the issuer have a better understanding of their current and potential customers,
which would inform their future strategy, including their planning of offering targeted credit
products to their customers.
The credit card issuer has gathered information on 30000 customers. The dataset contains
information on 24 variables, including demographic factors, credit data, history of payment, and
bill statements of credit card customers from April 2005 to September 2005, as well as
information on the outcome: did the customer default or not?
###Can we help VivaCredit to slove the problem?

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_csv('viva_credit.csv')

In [3]:
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
0   1    20000.0    2          2         1   24      2      2     -1     -1   
1   2   120000.0    2          2         2   26     -1      2      0      0   
2   3    90000.0    2          2         2   34      0      0      0      0   
3   4    50000.0    2          2         1   37      0      0      0      0   
4   5    50000.0    1          2         1   57     -1      0     -1      0   

   ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0  ...        0.0        0.0        0.0       0.0     689.0       0.0   
1  ...     3272.0     3455.0     3261.0       0.0    1000.0    1000.0   
2  ...    14331.0    14948.0    15549.0    1518.0    1500.0    1000.0   
3  ...    28314.0    28959.0    29547.0    2000.0    2019.0    1200.0   
4  ...    20940.0    19146.0    19131.0    2000.0   36681.0   10000.0   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  default.payment.next.month  
0       0.0       0.0       0.0                           1  
1    1000.0       0.0    2000.0                           1  
2    1000.0    1000.0    5000.0                           0  
3    1100.0    1069.0    1000.0                           0  
4    9000.0     689.0     679.0                           0  

[5 rows x 25 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ID                          30000 non-null  int64  
 1   LIMIT_BAL                   30000 non-null  float64
 2   SEX                         30000 non-null  int64  
 3   EDUCATION                   30000 non-null  int64  
 4   MARRIAGE                    30000 non-null  int64  
 5   AGE                         30000 non-null  int64  
 6   PAY_0                       30000 non-null  int64  
 7   PAY_2                       30000 non-null  int64  
 8   PAY_3                       30000 non-null  int64  
 9   PAY_4                       30000 non-null  int64  
 10  PAY_5                       30000 non-null  int64  
 11  PAY_6                       30000 non-null  int64  
 12  BILL_AMT1                   30000 non-null  float64
 13  BILL_AMT2                   300

In [5]:
df.describe()

ID       LIMIT_BAL           SEX     EDUCATION      MARRIAGE  \
count  30000.000000    30000.000000  30000.000000  30000.000000  30000.000000   
mean   15000.500000   167484.322667      1.603733      1.853133      1.551867   
std     8660.398374   129747.661567      0.489129      0.790349      0.521970   
min        1.000000    10000.000000      1.000000      0.000000      0.000000   
25%     7500.750000    50000.000000      1.000000      1.000000      1.000000   
50%    15000.500000   140000.000000      2.000000      2.000000      2.000000   
75%    22500.250000   240000.000000      2.000000      2.000000      2.000000   
max    30000.000000  1000000.000000      2.000000      6.000000      3.000000   

                AGE         PAY_0         PAY_2         PAY_3         PAY_4  \
count  30000.000000  30000.000000  30000.000000  30000.000000  30000.000000   
mean      35.485500     -0.016700     -0.133767     -0.166200     -0.220667   
std        9.217904      1.123802      1.197186      1.196868      1.169139   
min       21.000000     -2.000000     -2.000000     -2.000000     -2.000000   
25%       28.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%       34.000000      0.000000      0.000000      0.000000      0.000000   
75%       41.000000      0.000000      0.000000      0.000000      0.000000   
max       79.000000      8.000000      8.000000      8.000000      8.000000   

       ...      BILL_AMT4      BILL_AMT5      BILL_AMT6       PAY_AMT1  \
count  ...   30000.000000   30000.000000   30000.000000   30000.000000   
mean   ...   43262.948967   40311.400967   38871.760400    5663.580500   
std    ...   64332.856134   60797.155770   59554.107537   16563.280354   
min    ... -170000.000000  -81334.000000 -339603.000000       0.000000   
25%    ...    2326.750000    1763.000000    1256.000000    1000.000000   
50%    ...   19052.000000   18104.500000   17071.000000    2100.000000   
75%    ...   54506.000000   50190.500000   49198.250000    5006.000000   
max    ...  891586.000000  927171.000000  961664.000000  873552.000000   

           PAY_AMT2      PAY_AMT3       PAY_AMT4       PAY_AMT5  \
count  3.000000e+04   30000.00000   30000.000000   30000.000000   
mean   5.921163e+03    5225.68150    4826.076867    4799.387633   
std    2.304087e+04   17606.96147   15666.159744   15278.305679   
min    0.000000e+00       0.00000       0.000000       0.000000   
25%    8.330000e+02     390.00000     296.000000     252.500000   
50%    2.009000e+03    1800.00000    1500.000000    1500.000000   
75%    5.000000e+03    4505.00000    4013.250000    4031.500000   
max    1.684259e+06  896040.00000  621000.000000  426529.000000   

            PAY_AMT6  default.payment.next.month  
count   30000.000000                30000.000000  
mean     5215.502567                    0.221200  
std     17777.465775                    0.415062  
min         0.000000                    0.000000  
25%       117.750000                    0.000000  
50%      1500.000000                    0.000000  
75%      4000.000000                    0.000000  
max    528666.000000                    1.000000  

[8 rows x 25 columns]

In [6]:
df.isnull().sum()

ID                            0
LIMIT_BAL                     0
SEX                           0
EDUCATION                     0
MARRIAGE                      0
AGE                           0
PAY_0                         0
PAY_2                         0
PAY_3                         0
PAY_4                         0
PAY_5                         0
PAY_6                         0
BILL_AMT1                     0
BILL_AMT2                     0
BILL_AMT3                     0
BILL_AMT4                     0
BILL_AMT5                     0
BILL_AMT6                     0
PAY_AMT1                      0
PAY_AMT2                      0
PAY_AMT3                      0
PAY_AMT4                      0
PAY_AMT5                      0
PAY_AMT6                      0
default.payment.next.month    0
dtype: int64

In [7]:
X=df.drop('default.payment.next.month',axis=1)
y=df['default.payment.next.month']

In [8]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [9]:
from sklearn.linear_model import LogisticRegression

In [10]:
log_ref=LogisticRegression()

In [11]:
log_ref.fit(X_train,y_train)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [12]:
y_pred=log_ref.predict(X_test)
y_pred

array([0, 0, 0, ..., 0, 0, 0])

In [13]:
from sklearn.metrics import accuracy_score

In [14]:
accuracy_score(y_test,y_pred)

0.7811666666666667

In [15]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [16]:
log_ref_scaled = LogisticRegression()
log_ref_scaled.fit(X_train_scaled, y_train)

LogisticRegression()

In [17]:
y_pred_scaled = log_ref_scaled.predict(X_test_scaled)
accuracy_scaled = accuracy_score(y_test, y_pred_scaled)
print(f"Accuracy after scaling: {accuracy_scaled}")

Accuracy after scaling: 0.8098333333333333


In [18]:
from sklearn.metrics import precision_score, recall_score, f1_score, roc_auc_score

precision = precision_score(y_test, y_pred_scaled)
recall = recall_score(y_test, y_pred_scaled)
f1 = f1_score(y_test, y_pred_scaled)
auc_roc = roc_auc_score(y_test, log_ref_scaled.predict_proba(X_test_scaled)[:, 1])

print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1-score: {f1}")
print(f"AUC-ROC: {auc_roc}")

Precision: 0.6928251121076233
Recall: 0.23533891850723535
F1-score: 0.35133598635588403
AUC-ROC: 0.7270556485659562


##  Project Documentation

### 1.  Data Loading and Initial Exploration

- Loaded the dataset from `viva_credit.csv` using pandas.
- Displayed the first few rows with `df.head()` to understand the structure.
- Used `df.info()` to check data types and confirm the absence of null values.
- Generated statistical summaries using `df.describe()` to understand central tendencies and spread.
- Validated that there are no missing values using `df.isnull().sum()`.

---

### 2.  Data Preparation

- Separated features (X) and target (y) where the target is `'default.payment.next.month'`.
- Split the data into training and testing sets using `train_test_split(test_size=0.2, random_state=42)`.
- Applied `StandardScaler` to normalize features and enhance model convergence/performance.

---

### 3.  Model Training and Evaluation (Logistic Regression)

- Trained a Logistic Regression model on the unscaled training data.
  - Encountered a convergence warning, likely due to unnormalized input.
- Predicted outcomes on the test data and calculated accuracy.
- Retrained the Logistic Regression model using the scaled data.
- Made predictions again using the scaled test set.
- Observed improved accuracy after applying feature scaling.



## Summary

### Data Analysis Key Findings

- The scaled Logistic Regression model achieved:
  - Precision: 0.6928
  - Recall: 0.2353
  - F1-score: 0.3513
  - AUC-ROC: 0.7271

- The most influential predictor of default is `PAY_0` (repayment status in September 2005), based on the highest absolute coefficient in the Logistic Regression model.

- Other significant features impacting default include:
  - `BILL_AMT1`: Higher bill amount (September 2005) is associated with a higher likelihood of default.
  - `PAY_AMT1` and `PAY_AMT2`: Higher payment amounts (September and August 2005) are associated with a lower likelihood of default.

- Performance comparison of different models based on AUC-ROC:
  - Gradient Boosting: 0.7832 (best)
  - Random Forest: 0.7599
  - Logistic Regression: 0.7271

- Gradient Boosting also achieved:
  - Highest Accuracy: 0.8205
  - Highest F1-score: 0.4634

- While Logistic Regression had the best Precision (0.6928), Gradient Boosting provided the best overall balance across all performance metrics.
